server/testnet.go

```diff
@@ -0,0 +1,174 @@
+	return cmd
+}
+
+func testnetWithConfig(config *cfg.Config, ctx *Context, cdc *wire.Codec, appInit AppInit) error {
+
+	// Generate private key, node ID, initial transaction
+	for i := 0; i < viper.GetInt(nValidators); i++ {
+		nodeDirName := fmt.Sprintf("%s%d", viper.GetString(nodeDirPrefix), i)
+		nodeDir := filepath.Join(viper.GetString(outputDir), nodeDirName, "gaiad")
+		clientDir := filepath.Join(viper.GetString(outputDir), nodeDirName, "gaiacli")
+		gentxsDir := filepath.Join(viper.GetString(outputDir), "gentxs")
+		config.SetRoot(nodeDir)
+
+		err := os.MkdirAll(filepath.Join(nodeDir, "config"), nodeDirPerm)
+		if err != nil {
+			_ = os.RemoveAll(viper.GetString(outputDir))
+			return err
+		}
+
+		err = os.MkdirAll(clientDir, nodeDirPerm)
+		if err != nil {
+			_ = os.RemoveAll(viper.GetString(outputDir))
+			return err
+		}
+
+		config.Moniker = nodeDirName
+
+		ip := viper.GetString(startingIPAddress)
+		if len(ip) == 0 {
+			ip, err = externalIP()
+			if err != nil {
+				return err
+			}
+		} else {
+			ip = calculateIP(ip, i)
+		}
+
+		genTxConfig := gc.GenTxConfig{
+			nodeDirName,
+			clientDir,
+			true,
+			ip,
+		}
+
+		// Run `init gen-tx` and generate initial transactions
+		cliPrint, genTxFile, err := gentxWithConfig(ctx, cdc, appInit, config, genTxConfig)
+		if err != nil {
+			return err
+		}
+
+		// Save private key seed words
+		name := fmt.Sprintf("%v.json", "key_seed")
+		writePath := filepath.Join(clientDir)
+		file := filepath.Join(writePath, name)
+		err = cmn.EnsureDir(writePath, 0700)
+		if err != nil {
+			return err
+		}
+		err = cmn.WriteFile(file, cliPrint, 0600)
+		if err != nil {
+			return err
+		}
+
+		// Gather gentxs folder
+		name = fmt.Sprintf("%v.json", nodeDirName)
+		writePath = filepath.Join(gentxsDir)
+		file = filepath.Join(writePath, name)
+		err = cmn.EnsureDir(writePath, 0700)
+		if err != nil {
+			return err
+		}
+		err = cmn.WriteFile(file, genTxFile, 0644)
+		if err != nil {
+			return err
+		}
+
+	}
+
+	// Generate genesis.json and config.toml
+	chainID := "chain-" + cmn.RandStr(6)
+	for i := 0; i < viper.GetInt(nValidators); i++ {
+
+		nodeDirName := fmt.Sprintf("%s%d", viper.GetString(nodeDirPrefix), i)
+		nodeDir := filepath.Join(viper.GetString(outputDir), nodeDirName, "gaiad")
+		gentxsDir := filepath.Join(viper.GetString(outputDir), "gentxs")
+		gaiaInitConfig := InitConfig{
+			chainID,
+			true,
+			gentxsDir,
+			true,
+		}
+		config.Moniker = nodeDirName
+		config.SetRoot(nodeDir)
+
+		// Run `init` and generate genesis.json and config.toml
+		_, _, _, err := initWithConfig(ctx, cdc, appInit, config, gaiaInitConfig)
+		if err != nil {
+			return err
+		}
+	}
+
+	fmt.Printf("Successfully initialized %v node directories\n", viper.GetInt(nValidators))
+	return nil
+}
+
+func calculateIP(ip string, i int) string {
+	ipv4 := net.ParseIP(ip).To4()
+	if ipv4 == nil {
+		fmt.Printf("%v: non ipv4 address\n", ip)
+		os.Exit(1)
```

Comment:
Should pass be passing errors instead of just exiting here

Reply:
Yep, agreed. This was also copied from Tendermint as-is. I'll change it to pass an error instead.